File: api/schools/classes.py
from django.db.models import Count
from django.views import View
from django.http import JsonResponse
from django.views.generic import ListView
from django.views.decorators.csrf import csrf_exempt

from school.models import Sinf, Maktab, Belgisi
import json


@csrf_exempt
def add_class(request):
    """
    Add a new class (Sinf) associated with a school (Maktab) and optional badge (Belgisi).
    """
    if request.method == "POST":
        try:
            data = json.loads(request.body)  # Parse JSON data

            # Extract data from the request
            maktab_id = data.get("maktab_id")
            sinf_raqami = data.get("sinf_raqami")
            belgisi_nomi = data.get("belgisi")

            # Validate mandatory fields
            if not (maktab_id and sinf_raqami):
                return JsonResponse({"error": "Maktab ID va sinf raqami majburiy."}, status=400)

            # Get the related Maktab
            try:
                maktab = Maktab.objects.get(id=maktab_id)
            except Maktab.DoesNotExist:
                return JsonResponse({"error": "Maktab topilmadi."}, status=404)

            # Handle badge creation or assignment
            belgisi = None
            if belgisi_nomi:
                belgisi, _ = Belgisi.objects.get_or_create(nomi=belgisi_nomi.strip().upper())

            # Create the Sinf
            sinf = Sinf.objects.create(
                maktab=maktab,
                sinf_raqami=sinf_raqami.strip(),
                belgisi=belgisi
            )

            return JsonResponse({"message": "Sinf muvaffaqiyatli qo'shildi.", "id": sinf.id}, status=201)

        except json.JSONDecodeError:
            return JsonResponse({"error": "Invalid JSON format."}, status=400)
        except Exception as e:
            return JsonResponse({"error": f"Server xatosi: {str(e)}"}, status=500)

    return JsonResponse({"error": "Invalid request method."}, status=400)


class StatsClassAPIView(View):
    def get(self, request, *args, **kwargs):
        try:
            # Calculate statistics
            total_classes = Sinf.objects.count()
            total_schools = Maktab.objects.count()

            # Group classes by viloyat, tuman, and maktab
            grouped_data = (
                Sinf.objects
                .select_related('maktab', 'belgisi')
                .values('maktab__viloyat', 'maktab__tuman', 'maktab__id', 'maktab__nomi')
                .annotate(class_count=Count('id'))
                .order_by('maktab__viloyat', 'maktab__tuman', 'maktab__nomi')
            )

            # Structure the result
            grouped_result = {}
            for entry in grouped_data:
                viloyat = entry['maktab__viloyat']
                tuman = entry['maktab__tuman']
                maktab = {
                    "id": entry['maktab__id'],
                    "nomi": entry['maktab__nomi'],
                    "class_count": entry['class_count']
                }

                if viloyat not in grouped_result:
                    grouped_result[viloyat] = {}
                if tuman not in grouped_result[viloyat]:
                    grouped_result[viloyat][tuman] = []

                grouped_result[viloyat][tuman].append(maktab)

            return JsonResponse({
                'success': True,
                'stats': {
                    'total_classes': total_classes,
                    'total_schools': total_schools
                },
                'grouped_classes': grouped_result
            })

        except Exception as e:
            return JsonResponse({
                'success': False,
                'message': f"Xatolik yuz berdi: {str(e)}"
            }, status=500)

File: src/assets/customJs/schools/classes.js
document.addEventListener("DOMContentLoaded", () => {
  const regionSelect = document.getElementById("regionSelect");
  const districtSelect = document.getElementById("districtSelect");
  const schoolSelect = document.getElementById("schoolSelect");
  const addClassForm = document.getElementById("addClassForm");
  const groupedClassesDiv = document.getElementById("groupedClasses");
  const totalClassesElem = document.getElementById("totalClasses");
  const totalSchoolsElem = document.getElementById("totalSchools");

  const resetDropdown = (dropdown, placeholder) => {
    dropdown.innerHTML = `<option value="" disabled selected>${placeholder}</option>`;
    dropdown.disabled = true;
  };

  // Load statistics and update the UI
  const loadStatistics = async () => {
    try {
      const response = await fetch("/api/classes/stats/");
      const data = await response.json();

      if (data.success) {
        totalClassesElem.textContent = data.stats.total_classes || 0;
        totalSchoolsElem.textContent = data.stats.total_schools || 0;
      } else {
        toastr.error("Statistikani yuklashda xatolik yuz berdi.");
      }
    } catch (error) {
      toastr.error("Server bilan bog'lanishda xatolik yuz berdi.");
      console.error("Error loading statistics:", error);
    }
  };

  // Load grouped classes from API
  const loadGroupedClasses = async () => {
    try {
      const response = await fetch("/api/classes/stats/");
      const data = await response.json();

      if (data.success) {
        displayGroupedClasses(data.grouped_classes);
      } else {
        toastr.error("Ma'lumotlarni yuklashda xatolik yuz berdi.");
      }
    } catch (error) {
      toastr.error("Server bilan bog'lanishda xatolik yuz berdi.");
      console.error("Error loading grouped classes:", error);
    }
  };

  // Display grouped classes
  const displayGroupedClasses = (groupedData) => {
    groupedClassesDiv.innerHTML = "";

    for (const [viloyat, tumans] of Object.entries(groupedData)) {
      const regionDiv = document.createElement("div");
      regionDiv.className = "mt-4";
      regionDiv.innerHTML = `<h5 class="text-primary">${viloyat}</h5>`;

      for (const [tuman, schools] of Object.entries(tumans)) {
        const districtDiv = document.createElement("div");
        districtDiv.className = "mt-2";
        districtDiv.innerHTML = `<h6>${tuman}</h6>`;

        const schoolList = document.createElement("ul");
        schoolList.className = "list-group";

        schools.forEach((school) => {
          const schoolItem = document.createElement("li");
          schoolItem.className = "list-group-item";
          schoolItem.textContent = `${school.nomi} (${school.class_count} sinf)`;
          schoolList.appendChild(schoolItem);
        });

        districtDiv.appendChild(schoolList);
        regionDiv.appendChild(districtDiv);
      }

      groupedClassesDiv.appendChild(regionDiv);
    }
  };

  // Load regions (viloyatlar)
  const loadRegions = async () => {
    try {
      const response = await fetch("/api/schools/grouped/");
      const data = await response.json();

      if (data.data) {
        for (const viloyat in data.data) {
          const option = document.createElement("option");
          option.value = viloyat;
          option.textContent = viloyat;
          regionSelect.appendChild(option);
        }
        regionSelect.disabled = false;
      } else {
        toastr.error("Viloyatlar ma'lumotini yuklashda xatolik yuz berdi.");
      }
    } catch (error) {
      console.error("Error loading regions:", error);
      toastr.error("Server bilan bog'lanishda xatolik yuz berdi.");
    }
  };

  // Load districts (tumanlar) based on selected region
  regionSelect.addEventListener("change", () => {
    const selectedRegion = regionSelect.value;
    resetDropdown(districtSelect, "Tuman tanlang");
    resetDropdown(schoolSelect, "Maktabni tanlang");

    if (!selectedRegion) return;

    try {
      fetch("/api/schools/grouped/")
        .then((response) => response.json())
        .then((data) => {
          const districts = data.data[selectedRegion];
          if (districts) {
            for (const tuman in districts) {
              const option = document.createElement("option");
              option.value = tuman;
              option.textContent = tuman;
              districtSelect.appendChild(option);
            }
            districtSelect.disabled = false;
          } else {
            toastr.warning("Tanlangan viloyatda tumanlar topilmadi.");
          }
        });
    } catch (error) {
      console.error("Error loading districts:", error);
      toastr.error("Server bilan bog'lanishda xatolik yuz berdi.");
    }
  });

  // Load schools (maktablar) based on selected district
  districtSelect.addEventListener("change", () => {
    const selectedRegion = regionSelect.value;
    const selectedDistrict = districtSelect.value;
    resetDropdown(schoolSelect, "Maktabni tanlang");

    if (!selectedRegion || !selectedDistrict) return;

    try {
      fetch("/api/schools/grouped/")
        .then((response) => response.json())
        .then((data) => {
          const schools = data.data[selectedRegion][selectedDistrict]?.maktablar;
          if (schools) {
            schools.forEach((school) => {
              const option = document.createElement("option");
              option.value = school.id;
              option.textContent = `${school.maktab_raqami || "No Number"} | ${school.nomi}`;
              schoolSelect.appendChild(option);
            });
            schoolSelect.disabled = false;
          } else {
            toastr.warning("Tanlangan tumanda maktablar topilmadi.");
          }
        });
    } catch (error) {
      console.error("Error loading schools:", error);
      toastr.error("Server bilan bog'lanishda xatolik yuz berdi.");
    }
  });

  // Submit the add class form
  addClassForm.addEventListener("submit", async (e) => {
    e.preventDefault();

    const maktabId = schoolSelect.value;
    const sinfRaqami = document.getElementById("classNumber").value;
    const belgisi = document.getElementById("badgeName").value;

    if (!maktabId || !sinfRaqami) {
      toastr.error("Maktab va sinf raqami majburiy.");
      return;
    }

    const csrfToken = document.querySelector('[name=csrfmiddlewaretoken]').value;

    try {
      const response = await fetch("/api/classes/add/", {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "X-CSRFToken": csrfToken,
        },
        body: JSON.stringify({
          maktab_id: maktabId,
          sinf_raqami: sinfRaqami,
          belgisi: belgisi || null,
        }),
      });

      const result = await response.json();
      if (response.ok) {
        toastr.success(result.message || "Sinf muvaffaqiyatli qo'shildi.");
        addClassForm.reset();
        resetDropdown(districtSelect, "Tuman tanlang");
        resetDropdown(schoolSelect, "Maktabni tanlang");
        loadStatistics(); // Reload statistics after adding a new class
        loadGroupedClasses(); // Reload grouped classes after adding a new class
      } else {
        toastr.error(result.error || "Xatolik yuz berdi.");
      }
    } catch (error) {
      console.error("Error:", error);
      toastr.error("Server bilan bog'lanishda xatolik yuz berdi.");
    }
  });

  // Initialize dropdowns on page load
  resetDropdown(regionSelect, "Viloyatni tanlang");
  resetDropdown(districtSelect, "Tuman tanlang");
  resetDropdown(schoolSelect, "Maktabni tanlang");

  // Load initial data
  loadRegions();
  loadStatistics();
  loadGroupedClasses();
});
